packages/xl-ai/src/AIExtension.ts

```diff
@@ -0,0 +1,195 @@
+import {
+  BlockNoteEditor,
+  BlockNoteExtension,
+  UnreachableCaseError,
+} from "@blocknote/core";
+import { LanguageModel } from "ai";
+import { createStore } from "zustand/vanilla";
+import { PromptOrMessages, llm } from "./api";
+import { LLMRequestOptions } from "./api/streamTool/callLLMWithStreamTools";
+
+type AIPluginState = {
+  aiMenuBlockID: string | undefined;
+  aiMenuResponseStatus: "initial" | "generating" | "error" | "done";
+};
+
+// parameters that are shared across all calls and can be configured on the context as "application wide" settings
+type GlobalLLMCallOptions = {
+  model: LanguageModel;
+  dataFormat?: "html" | "json" | "markdown";
+  stream?: boolean;
+};
+
+// parameters that are specific to each call
+type CallSpecificCallLLMOptions = Partial<GlobalLLMCallOptions> &
+  Omit<LLMRequestOptions, "messages" | "model"> &
+  PromptOrMessages & {
+    defaultStreamTools?: {
+      /** Enable the add tool (default: true) */
+      add?: boolean;
+      /** Enable the update tool (default: true) */
+      update?: boolean;
+      /** Enable the delete tool (default: true) */
+      delete?: boolean;
+    };
+  };
+
+// TO DISCUSS: extension vs plugin vs addon vs ..
+export class AIExtension extends BlockNoteExtension {
+  public static name(): string {
+    return "ai";
+  }
+
+  /**
+   * zustand design considerations
+   *
+   * - setters are not added in this store, but as class methods so we can have control over what to expose as API
+   * - con: if we'd expose the store externally, consumers could still call store.setState :/ (for this store it's not desired, for `options` below, it is)
+   *
+   */
+  public readonly store = createStore<AIPluginState>()((set) => ({
+    aiMenuBlockID: undefined,
+    aiMenuResponseStatus: "initial",
+  }));
```

Comment:
Agree that setters being in the store would be weird, assuming that you have access to the class methods. If you didn't have access to the class methods, it would basically force all logic to be in zustand, which I think is the wrong way about things.

Being able to update state you don't own is pretty problematic 🙁  (if someone called store.setState), but I don't think we need to solve it or disable it.

Reply:
> Being able to update state you don't own is pretty problematic 🙁 (if someone called store.setState), but I don't think we need to solve it or disable it.

agree - as long as everywhere else we follow the pattern that we expose separate setters, I think we should be good